### PR TITLE
TRUNK-4747 Move Allergyapi into core

### PR DESCRIPTION
**Pull request by jdegraft**
Move of allergyapi to openmrs-core
- moved allergyapi-rest into module

### CODE REVIEW

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/AllergyResource1_9.java [hunk 1]
```diff
@@ -20,7 +20,7 @@
 import org.openmrs.module.webservices.rest.web.annotation.Resource;
 import org.openmrs.module.webservices.rest.web.v1_0.resource.openmrs1_8.AllergyResource1_8;
 
-@Resource(name = RestConstants.VERSION_1 + "/allergy", supportedClass = Allergy.class, supportedOpenmrsVersions = {"1.9.*", "1.10.*", "1.11.*", "1.12.*"})
+@Resource(name = RestConstants.VERSION_1 + "/allergy", order = 90, supportedClass = Allergy.class, supportedOpenmrsVersions = {"1.9.*", "1.10.*", "1.11.*", "1.12.*"})
```
Comment (rkorytkowski) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/AllergyResource1_9.java]: Why did you add order?

Comment (jdegraft) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/AllergyResource1_9.java]: This is due to some confusion I caused.
At the time, I mistakenly combined this task with the upgrade for Hibernate 4.0.
Darius suggested to drop support for 1.8 since 1.8 did not support DbSessionFactory, and I couldn't see at the time how to support that upgrade.
When you change the resources from 1.8 to 1.9 the conflict requires the order to resolve.
I guess I can redo the change and only leave the code to support the allergyapi change.
My understanding now indicates that this should be the recommended solution. The Hibernate upgrade should be a different commit and also there are better ways to support those sorts of changes.
What do you suggest?

File: omod-1.12/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/converter/openmrs1_12/AllergenConverter1_12.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import org.openmrs.module.webservices.rest.SimpleObject;
+import org.openmrs.module.webservices.rest.web.ConversionUtil;
+import org.openmrs.module.webservices.rest.web.representation.Representation;
+import org.openmrs.module.webservices.rest.web.response.ConversionException;
+
+/**
+ * Outputs a REST-compatible representation of an Allergen
+ */
+@Handler(supports = Allergen.class)
+public class AllergenConverter1_12 extends OutputOnlyConverter1_12<Allergen> {
```
Comment (rkorytkowski) [inline omod-1.12/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/converter/openmrs1_12/AllergenConverter1_12.java]: Please do not implement converters. They are not meant to be implemented directly in most cases. DataDelegatingCrudResource and MetadataDelegatingCrudResource already implement Converter and provides a basic implementation for asRepresentation. You should just implement getRepresentationDescription, getCreatableProperties. See AllergyResource1_8 for example.

File: omod-1.12/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/converter/openmrs1_12/AllergyConverter1_12.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import org.openmrs.module.webservices.rest.SimpleObject;
+import org.openmrs.module.webservices.rest.web.ConversionUtil;
+import org.openmrs.module.webservices.rest.web.representation.Representation;
+import org.openmrs.module.webservices.rest.web.response.ConversionException;
+
+/**
+ * Outputs a REST-compatible representation of a single Allergy (within an Allergies object)
+ */
+@Handler(supports = Allergy.class)
+public class AllergyConverter1_12 extends OutputOnlyConverter1_12<Allergy> {
```
Comment (rkorytkowski) [inline omod-1.12/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/converter/openmrs1_12/AllergyConverter1_12.java]: Same here, no need to implement AllergyConverter...

File: omod-1.12/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/converter/openmrs1_12/AllergyReactionConverter1_12.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import org.openmrs.module.webservices.rest.SimpleObject;
+import org.openmrs.module.webservices.rest.web.ConversionUtil;
+import org.openmrs.module.webservices.rest.web.representation.Representation;
+import org.openmrs.module.webservices.rest.web.response.ConversionException;
+
+/**
+ * Outputs a REST-compatible representation of an AllergyReaction
+ */
+@Handler(supports = AllergyReaction.class)
+public class AllergyReactionConverter1_12 extends OutputOnlyConverter1_12<AllergyReaction> {
```
Comment (rkorytkowski) [inline omod-1.12/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/converter/openmrs1_12/AllergyReactionConverter1_12.java]: Not needed.

File: omod-1.12/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_12/AllergiesResource1_12.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import org.openmrs.module.webservices.rest.web.RestConstants;
+import org.openmrs.module.webservices.rest.web.annotation.Resource;
+import org.openmrs.module.webservices.rest.web.representation.Representation;
+import org.openmrs.module.webservices.rest.web.resource.api.Retrievable;
+import org.openmrs.module.webservices.rest.web.response.ResponseException;
+
+import java.util.Arrays;
+import java.util.List;
+
+@Resource(name = RestConstants.VERSION_1 + "/allergies", supportedClass = Allergies.class, supportedOpenmrsVersions = {"1.9.*", "1.10.*", "1.11.*", "1.12.*"})
```
Comment (rkorytkowski) [inline omod-1.12/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_12/AllergiesResource1_12.java]: Could you please bring it up on talk.openmrs.org? I don't think it should be represented this way. It feels more like a subresource of a patient.

File: omod-1.12/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_12/BaseResource1_12Test.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.collection.IsIterableContainingInAnyOrder.containsInAnyOrder;
+import static org.hamcrest.core.Is.is;
+
+public abstract class BaseResource1_12Test extends BaseModuleWebContextSensitiveTest {
```
Comment (rkorytkowski) [inline omod-1.12/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_12/BaseResource1_12Test.java]: It feels that you should use MainResourceControllerTest instead, which seems to have all the features already. Any particular reason you added this class?

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AllergyResource2_0.java [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+		return super.doSearch(context);
+	}
+
+	/**
+	 * @see org.openmrs.module.webservices.rest.web.resource.impl.BaseDelegatingResource#getByUniqueId(java.lang.String)
+	 */
+	@SuppressWarnings("unchecked")
+	@Override
+	public Allergy getByUniqueId(String uniqueId) {
+		// TODO cannot be implemented because Allergies doesn't have a Patient
```
Comment (rkorytkowski) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_0/AllergyResource2_0.java]: You should call getPatientService.getAllergyByUuid here. I see that there's no such method in openmrs-core, but it must be simply added there.

File: omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AllergyController2_0Test.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+import org.openmrs.module.webservices.rest.SimpleObject;
+import org.openmrs.module.webservices.rest.test.Util;
+import org.openmrs.module.webservices.rest.web.RestConstants;
+import org.openmrs.module.webservices.rest.web.v1_0.resource.openmrs2_0.RestTestConstants2_0;
+import org.openmrs.module.webservices.rest.web.response.ResourceDoesNotSupportOperationException;
+import org.openmrs.module.webservices.rest.web.v1_0.controller.MainResourceControllerTest;
+import org.springframework.mock.web.MockHttpServletRequest;
+import org.springframework.web.context.request.WebRequest;
+
+// TODO: implement PatientService getAllergyByUuid
```
Comment (rkorytkowski) [inline omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/AllergyController2_0Test.java]: Just add that method to PatientService. It must be an oversight that it is missing.

File: omod-1.12/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_12/BaseResource1_12Test.java (hunk 6 above)
Comment (jdegraft) [inline omod-1.12/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_12/BaseResource1_12Test.java]: It seems this task is bigger than initially thought.
The idea was to move existing code from allergyapi with the least effort.
An initial assessment was the rest api aspect of allergyapi was not written in line with the way the current rest api was written.
Your review suggestions, indicate that this has to be completely revamped.
At the moment, I am porting the UI, allergyui, so it is difficult to see how a major rewrite here will impact the ui.

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/AllergyResource1_9.java (hunk 1 above)
Comment (rkorytkowski) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/AllergyResource1_9.java]: Just drop the order. It should not be needed.

File: omod-1.12/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_12/BaseResource1_12Test.java (hunk 6 above)
Comment (rkorytkowski) [inline omod-1.12/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_12/BaseResource1_12Test.java]: I see. I'm not familiar with the Allergy API and UI thus cannot make any assessments. I've just reviewed your changes from the webservices.rest module perspective and how we typically implement resources. I'll let others who worked on Allergy API to chime in.

**Comment by rkorytkowski** [comment]
@jdegraft, I've just commented on the issue that I'm probably not the best person to review your pull request so please hold on with any changes. It's been implemented in a way that is completely different from other webservices.rest resources and I cannot make informed decisions on the migration path.
